### PR TITLE
Fix RO apply link

### PR DESCRIPTION
**Pull request by notrichardpeng**
#7041 

In React, HTML written in locales break because JSX expressions automatically add double quotes to the translated sentences. This is fixed by moving the apply link to ruby views, but makes the link also appear in the case where there are no regional organizations, which on a second thought, makes sense too.

Case I'm talking about:

image

### CODE REVIEW

**Review by gregorbg — CHANGES_REQUESTED** [review]
I am not happy about backporting stuff to Ruby just because of minor nuisances like this.

Also, please consider: This is possible because the link text coincidentally appears at the bottom of the page, so you can extract it quite easily. What if link texts appear in the middle of the page instead? What about dynamic I18N texts?

Please research a "proper" React sokution for this kind of problem.

**Comment by notrichardpeng** [comment]
Great second point. After a while I found `dangerouslySetInnerHTML`, but I should include [DOMpurify](https://github.com/cure53/DOMPurify) for security purposes.

**Comment by gregorbg** [comment]
Thanks, in terms of your changes it's all good now!

However, I am still not happy about using `dangerouslySetInnerHTML` all the time. It will happen more and more as we start migrating more components to React, and it's definitely not good practice. I am open for ideas and debate how to improve this.

**Comment by coder13** [comment]
We could use markdown to react libraries and assert that the locales text is written in valid markdown

**Comment by gregorbg** [comment]
That would require us to either
1. Change our existing Ruby templates, which use the same i18n YAML
2. Maintain some keys in plain HTML template and some in Markdown, confusing both translators and developers.

I think the main issue is that Ruby and React have two different opinions about how to handle inline HTML in translations.

**Comment by gregorbg** [comment]
Separate YAML files are not an option because it makes life hard for non-tech-savvy translators.
What about creating a React component like `TranslationString` that can be used as 
```jsx
<SomeComponent foo="bar">
  <h3>Heading</h3>
  <TranslationString key="some.translation.yaml.key"/>
</SomeComponent>
```

It can then handle all of the `dompurify` and innerHTML stuff by itself and other parts of the code can reuse it as well.

**Comment by notrichardpeng** [comment]
Ohh okay, seems good. So at last we can trust `dangerouslySetInnerHTML` + `dompurify`?

**Comment by coder13** [comment]
I think it's fine to use `dangerouslySetInnerHTML` as long as you know what you're doing. That's why it's called `dangerouslySetInnerHTML`. We should make sure that our content is sanitized well as it goes into `dangerouslySetInnerHTML`.

**Comment by gregorbg** [comment]
+1 Caleb. I think this underlines the importance of creating a custom component, that can only be passed an i18n key and not actual HTML. In doing so, we don't encourage devs to `dangerouslySetInnerHTML` all the time and we also have implicit control over that part of sanitizing because the data only comes from one "source of truth" -- namely the translation YAMLs

**Comment by notrichardpeng** [comment]
Sounds like a viable solution. I will go ahead and implement this then?

**Comment by gregorbg** [comment]
Please do :)

**Comment by notrichardpeng** [comment]
I changed it to a more generic solution to avoid getting the "I18n key not used" test fail, but maybe this is actually just better because it can also be used for things other than translations?

**Comment by gregorbg** [comment]
At the current point in time, I am specifically opposed to allowing generic content here. We should limit it to i18n keys only.

Preventing the test fail is a matter of configuring our "find unused i18n keys" library in a way that detects these usages. See https://github.com/glebm/i18n-tasks/blob/main/templates/config/i18n-tasks.yml#L142 for reference.

**Review by gregorbg — APPROVED** [review]

**Comment by gregorbg** [comment]
Code is fine, now you just have to update the scanner for `i18n-tasks` to make sure it detects the usages.

**Comment by notrichardpeng** [comment]
I finally have more free time now. I looked at how to add the scanner again and this might take a while as I have no experience with regex :P ! Also `i18n-tasks` points to `custom_wca_i18n_scanner.rb`, are scanners for React also there? Want to make sure because I only see stuff for Ruby.

**Comment by gregorbg** [comment]
`CustomWcaI18NScanner` is its own thing that is not related to this feature. Just ignore it and pretend it doesn't even exist.

Essentially, `i18n-tasks` is doing a very stupid walk over all source files, looking for strings along the form of `t('some.key')` and if it finds such a string, it considers `'some.key'` as "used" and all is good.

You need to make the library understand that `<YourCoolNewReactComponent key='some.key' />` is equal to 't('some.key')' in terms of considering the i18n key as "used". To do so, you can use the pre-configured `PatternMapper` that comes delivered with the library, you just have to tell it which patterns to look out for.

The basic syntax is like this:
```ruby
I18n::Tasks.add_scanner 'I18n::Tasks::Scanners::PatternMapper',
    only: %w(*.html.haml *.html.slim),
    patterns: [['= title\b', '.page_title']] %>
```

where `only` is a simple glob pattern over the files you want to consider in the first place (probably something like `*.jsx` in this case) and `patterns` is an array of replacement rules where each replacement rule in itself is a two-element-array of the shape `[how_it_looks_in_our_actual_source_code, what_key_you_should_extract_from_it]`. One `PatternMapper` can hold arbitrarily many patterns, but in the above example it holds just one replacement pattern which makes the double array braces look weird.

Does that help?

**Comment by notrichardpeng** [comment]
Thanks Gregor that helped a lot! The tests are working locally and I just pushed the changes.

File: WcaOnRails/config/i18n-tasks.yml.erb [hunk 1]
```diff
@@ -154,6 +154,10 @@ ignore:
 <%
   require './lib/custom_wca_i18n_scanner.rb'
   I18n::Tasks.add_scanner 'CustomWcaI18nScanner', only: %w(*.erb)
+
+  I18n::Tasks.add_scanner 'I18n::Tasks::Scanners::PatternMapper',
+    only: %w(*.js),
+    patterns: [['I18nHTMLTranslate\si18nKey=%{key}', '%{key}']]
```
Comment (gregorbg) [inline WcaOnRails/config/i18n-tasks.yml.erb:160]: Please change the pattern to use something like `\s+` (where `+` is RegExp for "one or more") such that line breaks and tab characters aren't ignored by this pattern match.

**Review by gregorbg — CHANGES_REQUESTED** [review]
Cool! Small but important nit about the pattern you came up with. Everything else good to go!

**Review by gregorbg — APPROVED** [review]

**Pull request merged**